### PR TITLE
Add generic reflection over event names and versions

### PR DESCRIPTION
**Pull request by 50U10FCA7**
Requires #7

## Synopsis

At now, codegen machinery expands into non-generic code for handling `event::Name`s and `event::Revision`s.

## Solution

Provide two traits to add generic reflection:
- `event::codegen::Meta` for internal purposes (checking the name+ver collisions);
- `event::reflect::Meta` for public API, allowing to reflect over `event::Meta`s.

and one struct:
- `event::Meta` as a meta information of `Event` (`event::Name` and `event::Revision`)

## Checklist

- Created PR:
    - [x] In [draft mode][l:1]
    - [x] Name contains issue reference
    - [x] Has type and `k::` labels applied
    - [x] Has assignee
- Before [review][l:4]:
    - [x] Documentation is updated (if required)
    - [x] Tests are updated (if required)
    - [x] Changes conform [code style][l:2]
    - [x] [CHANGELOG entry][l:3] is added (if required)
    - [x] FCM (final commit message) is posted or updated
    - [x] [Draft mode][l:1] is removed
- [x] [Review][l:4] is completed and changes are approved
    - [x] FCM (final commit message) is approved
- Before merge:
    - [x] Milestone is set
    - [x] PR's name and description are correct and up-to-date
    - [x] All temporary labels are removed

[l:1]: https://help.github.com/en/articles/about-pull-requests#draft-pull-requests
[l:2]: /CONTRIBUTING.md#code-style
[l:3]: /CHANGELOG.md
[l:4]: https://help.github.com/en/articles/reviewing-changes-in-pull-requests

### CODE REVIEW

**Comment by 50U10FCA7** [comment]
### FCM
```
Add generic reflection over event names and versions (#6)

- add `es::event::reflect::Name` and `es::event::reflect::Revision` in `arcane-core` crate
- replace `es::event::codegen::Reflect` with `es::event::codegen::Meta` in `arcane-core` crate
```

File: core/src/es/event.rs [hunk 1]
```diff
@@ -215,21 +215,128 @@ impl<Ev: Event + ?Sized, S: Initialized<Ev>> Sourced<Initial<Ev>>
...
+    impl Iterator for VersionedNamesIterator {
+        type Item = (Name, Version);
+
+        fn next(&mut self) -> Option<Self::Item> {
+            self.0.next().map(|(_, name, ver)| {
+                // SAFETY: Safe, because `VersionedNamesIterator` can be
+                //         constructed only inside of this module and only from
+                //         `Events::EVENTS` which is guaranteed to contain valid
+                //         `Version`s (and requires `unsafe` to implement it).
+                let ver = unsafe { Version::new_unchecked(*ver) };
```
Comment (50U10FCA7) [inline core/src/es/event.rs:324]: @tyranron

**Comment by 50U10FCA7** [comment]
Discussed:

- add `es::event::reflect` module
- add `es::event::Meta { name: Name, revision: Revision }` struct
- add `es::event::reflect::Meta { const META: &'static [Meta] }` trait
- expand `es::event::reflect::Meta` in `es::Event`, `es::event::Revised` derive macros
- remove `es::event::VersionedNames`, `es::event::codegen::VersionedNamesIterator`

To discuss:

- namings

ack @tyranron

**Comment by 50U10FCA7** [comment]
@tyranron Meta is now represented as separated traits because we need to know the type of event revision (which requires `Revisable` bound).

**Review by tyranron — APPROVED** [review]

**Pull request merged**